drivers: video: improve support for custom controls

epc-ake opened the pull request:
# v4z: Custom control improvements

These are the changes I made to get my custom sensor running with the current API. I opened a draft PR for discussion.

- Adding name fields directly to `struct video_ctrl` to allow custom
  control names
- Adding `video_init_custom_ctrl()` function to initialize custom
 controls
- Make sure to allways set the right type for menu controls
~~- Skip new-value validation when `VIDEO_CTRL_FLAG_VOLATILE` is set. Consider adding an execute-on-write flag similar to `V4L2_CTRL_FLAG_EXECUTE_ON_WRITE`.~~
~~- Bug fixes in the video shell.~~

## Example usage
I got the inspiration from v4l2: 
https://www.kernel.org/doc/html/v4.10/media/kapi/v4l2-controls.html#custom-controls
Custom controls can now be initialized with the ´struct video_ctrl_config´.
```
static const struct video_ctrl_config example_int_cfg = {
    .id = EXAMPLE_CTRL_ID,
    .name = "Example Integer",
    .type = VIDEO_CTRL_TYPE_INTEGER,
    .flags = 0,
    .range = {.min = 0, .max = 100000, .step = 1, .def  = 100},
};
ret = video_init_custom_ctrl(&ctrls->example_int, dev, &example_int_cfg);
```

and for custom menu controls
```
static const char *const example_menu[] = {
    [0] = "Option A",
    [1] = "Option B",
    [2] = "Option C",
    NULL,
};

static const struct video_ctrl_config example_ctrl_cfg = {
    .id = EXAMPLE_CTRL_ID,
    .name = "Example Control",
    .type = VIDEO_CTRL_TYPE_MENU,
    .flags = 0,
    .menu = example_menu,
    .range = {.min = 0, .max = 2, .step = 1, .def  = 0},
};

ret |= video_init_custom_ctrl(&ctrls->example, dev, &example_ctrl_cfg);
```

ngphibang commented on drivers/video/video_ctrls.c:
```diff
@@ -589,7 +597,7 @@ int video_query_ctrl(struct video_ctrl_query *cq)
 		cq->int_menu = ctrl->int_menu;
 	}
-	cq->name = video_get_ctrl_name(cq->id);
+	cq->name = ctrl->name;
```
Suggested:

```
cq->name = video_get_ctrl_name(cq->id);
if (cq->name == NULL) {
    cq->name = ctrl->name;
}
```

ngphibang commented on drivers/video/video_ctrls.c:
```diff
@@ -156,6 +158,7 @@ int video_init_ctrl(struct video_ctrl *ctrl, const struct device *dev, uint32_t
 	ctrl->type = type;
 	ctrl->flags = flags;
 	ctrl->range = range;
+	ctrl->name = video_get_ctrl_name(id);
```
You need a new `video_init_ctrl_custom_ctrl(..., const char *name)` so that drivers can set the name

ngphibang commented on drivers/video/video_ctrls.c:
```diff
@@ -363,8 +368,11 @@ int video_set_ctrl(const struct device *dev, struct video_control *control)
 	/* No new value */
-	if (ctrl->type == VIDEO_CTRL_TYPE_INTEGER64 ? ctrl->val64 == control->val64
-						    : ctrl->val == control->val) {
+	if (!(ctrl->flags & VIDEO_CTRL_FLAG_VOLATILE) &&
```
VIDEO_CTRL_FLAG_EXECUTE_ON_WRITE is the right way to go I think. I can add it and the control button type if you want

epc-ake committed:
drivers: video: add support for custom controls

- Adding name fields directly to `struct video_ctrl` to allow custom
  control names
- Adding `video_init_custom_ctrl` to initialize custom
 controls
- Make sure to allways set the right type for menu controls

Signed-off-by: Armin Kessler <ake@espros.com>

epc-ake committed:
drivers: video: refactor video_ctrl

refactoring `struct video_ctrl` to make use of
`struct video_ctrl_config` to avoid duplications

Signed-off-by: Armin Kessler ake@espros.com

ngphibang commented:
@epc-ake : https://github.com/zephyrproject-rtos/zephyr/pull/95968 (added VIDEO_CTRL_FLAG_EXECUTE_ON_WRITE and VIDEO_CTRL_TYPE_BUTTON)

ngphibang commented on drivers/video/video_ctrls.h:
```diff
@@ -37,6 +39,19 @@ enum video_ctrl_type {
 
 struct video_device;
 
+struct video_ctrl_config {
+	uint32_t id;
+	const char *name;
+	enum video_ctrl_type type;
+	unsigned long flags;
+	struct video_ctrl_range range;
+	union {
+		const char *const *menu;
+		const int64_t *int_menu;
+	};
+	size_t menu_len;
+};
```
> However, when we want to use the default video_init_ctrl api's for custom controls we would have to set name, flags and type manually after every initialization.

- Even with standard control, drivers still need to explicitly set the `flags` and `type` if required, e.g. depending on HW, sometimes a control could be "volatile", sometimes not, etc.
- Driver holds  `struct video_ctrl` which is static so it can directly set its fields or passing parameters to `video_init_ctrl()` so that the framework can set it. The former approach is better in terms of performance (no need to pass parameters by copy) while the latter approach is a bit more "intuitive" (force developers to pass certain parameters cid and range, but again, they still set other things like flags and type if required).
- I 'd like to refactor video_init_ctrl() to something as below:


```
struct ov5640_ctrls {
        ...
	struct video_ctrl brightness = {.id = VIDEO_CID_BRIGHTNESS, . range = {.min = -15, .max = 15, .step = 1, .def = 0} };
	struct video_ctrl hflip = {.id = VIDEO_CID_HFLIP, .range = {.min = 0, .max = 1, .step = 1, .def = 0} };
};

video_init_ctrl(struct video_ctrl *ctrl, const struct device *dev)
```

This way, we drivers can specify any additional fields they feel needed, e.g. flag, type, name:

`struct video_ctrl custom = {.id = VIDEO_CID_CUSTOM, .range = {.min = 0, .max = 1, .step = 1, .def = 0}, .name = "custom", . type = VIDEO_CTRL_TYPE_BUTTON};`

and we have a consistent way to declare and init the controls for both standard and custom controls. We save some structs (video_ctrl_config), API (video_init_custom_ctrl) and some stack size (to pass parameters to functions)

@epc-ake @josuah @avolmat-st   What do you think ?

josuah replied:
This sounds like an intuitive way to solve this.
Controls looks a lot more like data than code, which leads to fewer code in the sensor.

This could need to be in the per-instance macro block:

```c
	.ctrls = { \
		.brightness = {.id = VIDEO_CID_BRIGHTNESS, . range = {.min = -15, .max = 15, .step = 1, .def = 0} }; \
		.hflip = {.id = VIDEO_CID_HFLIP, .range = {.min = 0, .max = 1, .step = 1, .def = 0} }; \
	}; \
```

For instance like found here:

https://github.com/zephyrproject-rtos/zephyr/blob/67f246477f93ac8e7bd4b953a9f47d4f5f5b4ef5/drivers/video/imx335.c#L511-L518

ngphibang replied:
yes, `per-instance` is important. This is the current limitation of the video control framework where controls are stored "per-driver". I will try to fix this at the end of  the week.

epc-ake replied:
Perfect. So are you taking over this? I will close this PR then.

ngphibang commented:
I continued this on https://github.com/zephyrproject-rtos/zephyr/pull/96393
